File: common/src/main/java/dev/matthiesen/cobble_paste/common/services/PokePastePartyProvider.java
package dev.matthiesen.cobble_paste.common.services;

import com.cobblemon.mod.common.api.npc.NPCPartyProvider;
import com.cobblemon.mod.common.api.storage.party.NPCPartyStore;
import com.cobblemon.mod.common.entity.npc.NPCEntity;
import com.cobblemon.mod.common.pokemon.Pokemon;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import dev.matthiesen.cobble_paste.common.api.PokePasteApiClient;
import dev.matthiesen.cobble_paste.common.formats.ShowdownTeam;
import dev.matthiesen.cobble_paste.common.util.PartyCache;
import kotlin.jvm.functions.Function1;
import net.minecraft.server.level.ServerPlayer;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public final class PokePastePartyProvider implements NPCPartyProvider {
    public static final String TYPE = "cobble_paste:poke_paste";
    public static Function1<String, NPCPartyProvider> PROVIDER = (id) -> new PokePastePartyProvider();

    private static final PartyCache CACHE = new PartyCache();

    public static void register() {
        NPCPartyProvider.Companion.getTypes().put(TYPE, PROVIDER);
    }

    private boolean isStatic = true;
    private String pokePasteId = "";


    @Override
    public @NotNull String getType() {
        return TYPE;
    }

    @Override
    public boolean isStatic() {
        return isStatic;
    }

    @Override
    public void loadFromJSON(@NotNull JsonElement jsonElement) {
        JsonObject object = jsonElement.getAsJsonObject();

        if (object.has("isStatic")) {
            isStatic = object.get("isStatic").getAsBoolean();
        }
        if (object.has("pokePasteId")) {
            pokePasteId = object.get("pokePasteId").getAsString();
        }
    }

    @Override
    public @NotNull NPCPartyStore provide(@NotNull NPCEntity npcEntity, int i, @NotNull List<? extends ServerPlayer> list) {
        List<Pokemon> party = this.loadTeam(i);
        if (party.isEmpty()) {
            NPCPartyStore empty = new NPCPartyStore(npcEntity);
            empty.initialize();
            return empty;
        }

        NPCPartyStore partyStore = new NPCPartyStore(npcEntity);
        int count = 0;

        for (Pokemon pokemon : party) {
            if (count >= 6) {
                break;
            }

            partyStore.add(pokemon);
            ++count;
        }

        partyStore.initialize();
        return partyStore;
    }

    private List<Pokemon> loadTeam(int level) {
        String cacheKey = pokePasteId + "_" + level;

        if (this.isStatic) {
            List<Pokemon> cachedParty = CACHE.get(cacheKey);
            if (cachedParty != null) {
                return cachedParty;
            }
        }

        String rawPaste = PokePasteApiClient.fetchRawPaste(pokePasteId).join();
        var showdownTeam = ShowdownTeam.fromPokePaste(rawPaste, PokePasteApiClient.extractPasteId(pokePasteId));
        List<Pokemon> party = new ArrayList<>();

        for (var entry : showdownTeam.team()) {
            if (party.size() >= 6) {
                break;
            }

            Pokemon pokemon = entry.toPokemon();
            if (entry.level().isEmpty()) {
                pokemon.setLevel(level);
            }
            party.add(pokemon);
        }

        List<Pokemon> loadedParty = List.copyOf(party);
        if (this.isStatic) {
            CACHE.put(cacheKey, loadedParty);
        }
        return loadedParty;
    }
}

File: common/src/main/kotlin/dev/matthiesen/cobble_paste/services/PokePastePartyProvider.kt
package dev.matthiesen.cobble_paste.services

import com.cobblemon.mod.common.api.npc.NPCPartyProvider
import com.cobblemon.mod.common.api.storage.party.NPCPartyStore
import com.cobblemon.mod.common.entity.npc.NPCEntity
import com.cobblemon.mod.common.pokemon.Pokemon
import com.google.gson.JsonElement
import dev.matthiesen.cobble_paste.common.CobblePasteCommon
import dev.matthiesen.cobble_paste.common.api.PokePasteApiClient
import dev.matthiesen.cobble_paste.common.formats.ShowdownTeam
import dev.matthiesen.cobble_paste.common.util.PartyCache
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerPlayer

/**
 * NPC Party provider that fetches a team from PokePaste and provides it to an NPC.
 * * This is written in Kotlin because the Cobblemon API is written in Kotlin, and it is easier to work with for some things.
 */
class PokePastePartyProvider : NPCPartyProvider {
    companion object {
        val TYPE: ResourceLocation = CobblePasteCommon.modResource("pokepaste")
        val CACHE: PartyCache = PartyCache()

        fun register() {
            NPCPartyProvider.types.putIfAbsent(TYPE, ::PokePastePartyProvider)
            CobblePasteCommon.INSTANCE.createInfoLog("Registered NPC PokePastePartyProvider")
        }
    }

    override val type = TYPE
    override var isStatic = true

    var pokePasteId = ""

    override fun loadFromJSON(json: JsonElement) {
        val jsonObject = json.asJsonObject

        if (jsonObject.has("isStatic")) {
            isStatic = jsonObject.get("isStatic").asBoolean
        }
        if (jsonObject.has("pokePasteId")) {
            pokePasteId = jsonObject.get("pokePasteId").asString
        }
    }

    override fun provide(
        npc: NPCEntity,
        level: Int,
        players: List<ServerPlayer>
    ): NPCPartyStore {
        val party = this.loadTeam(level)
        if (party.isEmpty()) {
            val emptyParty = NPCPartyStore(npc)
            emptyParty.initialize()
            return emptyParty
        }

        val partyStore = NPCPartyStore(npc)
        var count = 0

        party.forEach {
            if (count >= 6) return@forEach
            partyStore.add(it)
            count++
        }
        partyStore.initialize()
        return partyStore
    }

    fun loadTeam(level: Int): List<Pokemon> {
        val cacheKey = "${pokePasteId}_$level"

        if (this.isStatic) {
            val cachedParty = CACHE.get(cacheKey)
            if (cachedParty != null) {
                return cachedParty
            }
        }

        val rawPaste = PokePasteApiClient.fetchRawPaste(pokePasteId).join()
        val showdownTeam = ShowdownTeam.fromPokePaste(rawPaste, PokePasteApiClient.extractPasteId(pokePasteId))
        val party = ArrayList<Pokemon>()

        showdownTeam.team.forEach {
            if (party.size >= 6) return@forEach
            val pokemon = it.toPokemon()
            if (it.level().isEmpty) {
                pokemon.level = level
            }
            party.add(pokemon)
        }

        val loadedParty = party.toList()
        if (this.isStatic) {
            CACHE.put(cacheKey, loadedParty)
        }

        return loadedParty
    }
}